lib/Sema/TypeCheckProtocolInference.cpp

```diff
@@ -896,27 +854,28 @@ AssociatedTypeInference::computeDerivedTypeWitness(
   return result;
 }
 
-Type
+Optional<AbstractTypeWitness>
 AssociatedTypeInference::computeAbstractTypeWitness(
-                                              AssociatedTypeDecl *assocType) {
+    AssociatedTypeDecl *assocType) {
   // We don't have a type witness for this associated type, so go
   // looking for more options.
   if (Type concreteType = computeFixedTypeWitness(assocType))
-    return concreteType;
+    return AbstractTypeWitness{assocType, concreteType, nullptr};
 
   // If we can form a default type, do so.
-  if (Type defaultType = computeDefaultTypeWitness(assocType))
-    return defaultType;
+  if (auto typeWitness = computeDefaultTypeWitness(assocType))
+    return typeWitness;
 
   // If there is a generic parameter of the named type, use that.
   if (auto genericSig = dc->getGenericSignatureOfContext()) {
     for (auto gp : genericSig->getInnermostGenericParams()) {
       if (gp->getName() == assocType->getName())
-        return dc->mapTypeIntoContext(gp);
+        return AbstractTypeWitness{assocType, dc->mapTypeIntoContext(gp),
```

Comment:
I know it was already like that when you got here, but I think the mapTypeIntoContext() is probably unnecessary, if the caller just does mapTypeOutOfContext() again. At the very least, it's confusing because we don't mapTypeIntoContext() inside computeFixedTypeWitness() or computeDefaultTypeWitness() it seems.

Reply:
I actually had this idea as well and then abandoned it, assuming `mapTypeIntoContext` would save us some time by skipping substitutions in `completeSolution`, which are gated by `type->hasTypeParameter()`. But I guess the difference is disparaging, and I agree it looks out of place.

Reply:
So we end up always treating a generic parameter at index and depth 0 as Self during substitutions because of `type->isEqual(proto->getSelfInterfaceType())`, and everything goes wrong. We can keep the special case to skip substitutions or change the `TypeSubstitutionFn` to something like:
```cpp
[&](SubstitutableType *type) {
  if (auto gp = dyn_cast<GenericTypeParamType>(type)) {
    if (gp->getDecl() &&
        isa<ProtocolDecl>(gp->getDecl()->getDeclContext()))
    return adoptee;
  }

  return Type(type);
}
```

What do you prefer?

Reply:
Argh, the lamdba I wrote above is not an option either, because fixed type witnesses use canonical generic parameters e.g. `Array<τ_0_0.A>`, without a bound declaration.

Reply:
> if the caller just does mapTypeOutOfContext() again

I just noticed this. The caller doesn't `mapTypeOutOfContext()`, it only maps a type witness *into* context after we substitute. If you are keen to remove that call, we could add a `AbstractTypeWitnessKind` enum field and just skip the substitutions for the generic parameter case, but I don't feel confident about this. What do you think?